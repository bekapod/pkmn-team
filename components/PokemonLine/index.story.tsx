// tslint:disable-next-line:no-implicit-dependencies
import { storiesOf } from "@storybook/react";
import React from "react";
import PokemonLine from ".";

storiesOf("PokemonLine", module)
  .add("default", () => (
    <PokemonLine
      pokemon={{
        id: "25",
        name: "pikachu",
        pokedexId: 25,
        sprite: "25.png",
        types: [{ name: "ELECTRIC" }]
      }}
    />
  ))
  .add("with outdent", () => (
    <PokemonLine
      pokemon={{
        id: "25",
        name: "pikachu",
        pokedexId: 25,
        sprite: "25.png",
        types: [{ name: "ELECTRIC" }]
      }}
      outdent={10}
    />
  ));
