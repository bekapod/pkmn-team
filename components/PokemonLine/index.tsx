import React, { PureComponent } from "react";
import styled from "styled-components/macro";
import { formatPokemonName, sortTypes } from "../../helpers/general";
import { getTypeGradient } from "../../helpers/gradients";
import * as variables from "../../helpers/variables";
import { Pokemon, Type } from "../../types"; // eslint-disable-line import/named
import InlineList from "../InlineList";
import TypeTag from "../TypeTag";
import { media } from "../../helpers/media";

interface Props {
  pokemon: Pokemon;
  outdent: number;
  compact: boolean;
}

interface RowProps {
  outdent: number;
  compact: boolean;
  types: Type[];
}

const Row = styled.div`
  position: relative;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: ${variables.spacing.xxl}px;

  &::before {
    content: "";
    position: absolute;
    top: calc(${variables.spacing.sm}px / 2 * -1);
    left: ${({ outdent }: RowProps) => `-${outdent}px`};
    display: block;
    width: ${({ outdent }: RowProps) => `calc(100% + (${outdent}px * 2))`};
    height: ${variables.spacing.sm}px;
    background-image: ${({ types }: RowProps) => getTypeGradient(types)};
  }

  ${media.mediumOnly`
    ${({ compact }: RowProps) =>
      compact
        ? `
          flex-direction: column;
          padding: ${variables.spacing.md}px;
          height: ${variables.spacing.lg * 6}px;
          align-items: flex-start;
        `
        : ""}
  `}
`;

const RowImage = styled.img`
  width: ${variables.spacing.xl}px;
  height: ${variables.spacing.xl}px;
  margin-right: ${variables.spacing.md}px;

  ${media.mediumOnly`
    ${({ compact }: RowProps) =>
      compact
        ? `
        margin-right: 0;
        align-self: center;
      `
        : ""}
  `}
`;

const RowTitle = styled.div`
  margin-bottom: ${variables.spacing.xs}px;
  font-weight: 700;
  line-height: 1;
`;

class PokemonLine extends PureComponent<Props> {
  public static defaultProps = {
    outdent: 0,
    compact: false,
    pokemon: {}
  };

  public render(): JSX.Element {
    const { pokemon, compact, ...props } = this.props;
    const { pokedexId, name, types, sprite } = pokemon;

    return (
      <Row types={types} compact={compact} {...props}>
        <RowImage
          src={`/static/sprites/${sprite}`}
          alt={`${name} sprite`}
          compact={compact}
        />
        <div>
          <RowTitle>{formatPokemonName(pokemon)}</RowTitle>
          <InlineList>
            {sortTypes(types).map((type: Type) => (
              <li key={`Pokemon: ${pokedexId}, Type: ${type}`}>
                <TypeTag type={type}>{type}</TypeTag>
              </li>
            ))}
          </InlineList>
        </div>
      </Row>
    );
  }
}

export default PokemonLine;
