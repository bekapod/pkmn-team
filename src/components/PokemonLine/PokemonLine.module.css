.row {
  --outdent: 0px;

  position: relative;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: var(--spacing-10);

  &::before {
    content: '';
    position: absolute;
    top: calc(var(--spacing-2) / 2 * -1);
    left: calc(var(--outdent) * -1);
    display: block;
    width: calc(100% + (var(--outdent) * 2));
    height: var(--spacing-2);
    background-image: var(--type-gradient);
  }
}

.sprite {
  width: var(--spacing-8);
  height: var(--spacing-8);
  margin-right: var(--spacing-4);
}

.title {
  margin-bottom: var(--spacing-2);
  font-weight: 700;
  line-height: 1;
}
