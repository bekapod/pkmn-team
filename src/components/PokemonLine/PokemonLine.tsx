import type { ComponentPropsWithRef, FunctionComponent } from 'react';
import classNames from 'classnames';
import type { PokemonFragment, PokemonTypeFragment } from '~/generated/graphql';
import {
  formatPokemonName,
  getPokemonSpriteUrl,
  sortBySlug
} from '~/lib/general';
import { getTypeGradient } from '~/lib/gradients';
import { InlineList } from '../InlineList';
import { TypeTag } from '../TypeTag';

export type PokemonLineProps = {
  pokemon: PokemonFragment;
  outdent?: string;
};

export const PokemonLine: FunctionComponent<
  ComponentPropsWithRef<'div'> & PokemonLineProps
> = ({ pokemon, outdent, style, className, ...props }) => {
  const { pokedexId, name, types, sprite } = pokemon;

  return (
    <div
      className={classNames(
        'pokemon-line-template',
        'relative',
        'flex',
        'items-center',
        'h-10',
        className
      )}
      style={{
        ...style,
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        '--outdent': outdent,
        '--type-gradient': getTypeGradient(
          types.edges
            ?.map(edge => edge?.node)
            ?.filter((node): node is PokemonTypeFragment => !!node) ?? []
        )
      }}
      {...props}
    >
      <img
        className={classNames('w-8', 'h-8', 'mr-4')}
        src={getPokemonSpriteUrl(sprite)}
        alt={`${name} sprite`}
      />
      <div>
        <div className={classNames('mb-2', 'font-bold', 'leading-none')}>
          {formatPokemonName(pokemon)}
        </div>
        <InlineList>
          {sortBySlug(
            types.edges
              ?.map(edge => edge?.node)
              ?.filter((node): node is PokemonTypeFragment => !!node) ?? []
          ).map(type => (
            <li key={`Pokemon: ${pokedexId}, Type: ${type.slug}`}>
              <TypeTag typeSlug={type.slug}>{type?.name}</TypeTag>
            </li>
          ))}
        </InlineList>
      </div>
    </div>
  );
};
