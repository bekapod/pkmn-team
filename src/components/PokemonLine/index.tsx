import React, { PureComponent } from "react";
import styled from "styled-components/macro";
import { formatPokemonName, sortTypes } from "../../helpers/general";
import { getTypeGradient } from "../../helpers/gradients";
import * as variables from "../../helpers/variables";
import { IPokemon, Type } from "../../types";
import InlineList from "../InlineList";
import TypeTag from "../TypeTag";

interface IProps {
  pokemon: IPokemon;
  outdent: number;
}

interface IRowProps {
  outdent: number;
  types: Type[];
}

const Row = styled.div`
  position: relative;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: ${variables.spacing.xxl}px;

  &::before {
    content: "";
    position: absolute;
    top: calc(${variables.spacing.sm}px / 2 * -1);
    left: ${({ outdent }: IRowProps) => `-${outdent}px`};
    display: block;
    width: ${({ outdent }: IRowProps) => `calc(100% + (${outdent}px * 2))`};
    height: ${variables.spacing.sm}px;
    background-image: ${({ types }: IRowProps) => getTypeGradient(types)};
  }
`;

const RowImage = styled.img`
  width: ${variables.spacing.xl}px;
  height: ${variables.spacing.xl}px;
  margin-right: ${variables.spacing.md}px;
`;

const RowTitle = styled.div`
  font-weight: 700;
`;

class PokemonLine extends PureComponent<IProps> {
  public static defaultProps = {
    outdent: 0,
    pokemon: {}
  };

  public render() {
    const { pokemon, outdent } = this.props;
    const { pokedexId, name, types = [], sprite } = pokemon;

    return (
      <Row types={types} outdent={outdent}>
        <RowImage src={`/sprites/${sprite}`} alt={`${name} sprite`} />
        <div>
          <RowTitle>{formatPokemonName(pokemon)}</RowTitle>
          <InlineList>
            {sortTypes(types).map((type: Type) => (
              <li key={`Pokemon: ${pokedexId}, Type: ${type}`}>
                <TypeTag type={type}>{type}</TypeTag>
              </li>
            ))}
          </InlineList>
        </div>
      </Row>
    );
  }
}

export default PokemonLine;
