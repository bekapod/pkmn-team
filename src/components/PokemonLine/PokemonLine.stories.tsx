import { Meta, Story } from '@storybook/react/types-6-0';
import { pikachu } from '~/mocks/Pokemon';
import { PokemonLine, PokemonLineProps } from './PokemonLine';

export default {
  title: 'Components/PokemonLine',
  component: PokemonLine,
  args: {
    pokemon: pikachu
  }
} as Meta<PokemonLineProps>;

export const Standard: Story<PokemonLineProps> = args => (
  <PokemonLine {...args} />
);

export const WithOutdent: Story<PokemonLineProps> = args => (
  <PokemonLine {...args} />
);
WithOutdent.args = {
  outdent: '1rem'
};
