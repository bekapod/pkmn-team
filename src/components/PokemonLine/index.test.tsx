import React from "react";
import renderer from "react-test-renderer";
import PokemonLine from ".";

describe("<PokemonLine />", () => {
  it("renders without crashing", () => {
    const tree = renderer.create(
      <PokemonLine
        pokemon={{
          id: "25",
          name: "pikachu",
          pokedexId: 25,
          sprite: "25.png",
          types: ["ELECTRIC"]
        }}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  describe("when outdent is passed", () => {
    it("renders without crashing", () => {
      const tree = renderer.create(
        <PokemonLine
          pokemon={{
            id: "25",
            name: "pikachu",
            pokedexId: 25,
            sprite: "25.png",
            types: ["ELECTRIC"]
          }}
          outdent={10}
        />
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe("when a pokemon with more than one type is passed", () => {
    it("renders all types", () => {
      const tree = renderer.create(
        <PokemonLine
          pokemon={{
            id: "25",
            name: "bulbasaur",
            pokedexId: 1,
            sprite: "1.png",
            types: ["POISON", "GRASS"]
          }}
        />
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
