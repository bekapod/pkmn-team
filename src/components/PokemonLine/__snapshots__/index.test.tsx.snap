// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonLine /> renders without crashing 1`] = `
<div
  className="PokemonLine__Row-sc-1qiwepa-0 bfuspP"
>
  <img
    alt="pikachu sprite"
    className="PokemonLine__RowImage-sc-1qiwepa-1 hUcFnr"
    src="/sprites/25.png"
  />
  <div>
    <div
      className="PokemonLine__RowTitle-sc-1qiwepa-2 edSfRl"
    >
      #25 Pikachu
    </div>
    <ul
      className="InlineList-sc-1dn69qz-0 YzWxJ"
    >
      <li>
        <span
          className="TypeTag-sc-1dttu8y-0 byzsso"
          type="ELECTRIC"
        >
          ELECTRIC
        </span>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<PokemonLine /> when a pokemon with more than one type is passed renders all types 1`] = `
<div
  className="PokemonLine__Row-sc-1qiwepa-0 jeunMf"
>
  <img
    alt="bulbasaur sprite"
    className="PokemonLine__RowImage-sc-1qiwepa-1 hUcFnr"
    src="/sprites/1.png"
  />
  <div>
    <div
      className="PokemonLine__RowTitle-sc-1qiwepa-2 edSfRl"
    >
      #1 Bulbasaur
    </div>
    <ul
      className="InlineList-sc-1dn69qz-0 YzWxJ"
    >
      <li>
        <span
          className="TypeTag-sc-1dttu8y-0 jHkiSl"
          type="GRASS"
        >
          GRASS
        </span>
      </li>
      <li>
        <span
          className="TypeTag-sc-1dttu8y-0 jjNEEF"
          type="POISON"
        >
          POISON
        </span>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<PokemonLine /> when outdent is passed renders without crashing 1`] = `
<div
  className="PokemonLine__Row-sc-1qiwepa-0 iQKIEU"
>
  <img
    alt="pikachu sprite"
    className="PokemonLine__RowImage-sc-1qiwepa-1 hUcFnr"
    src="/sprites/25.png"
  />
  <div>
    <div
      className="PokemonLine__RowTitle-sc-1qiwepa-2 edSfRl"
    >
      #25 Pikachu
    </div>
    <ul
      className="InlineList-sc-1dn69qz-0 YzWxJ"
    >
      <li>
        <span
          className="TypeTag-sc-1dttu8y-0 byzsso"
          type="ELECTRIC"
        >
          ELECTRIC
        </span>
      </li>
    </ul>
  </div>
</div>
`;
